packages/flutter_tools/lib/src/web/asset_server.dart

```diff
@@ -0,0 +1,186 @@
+///     resolved relative to the dart tool directory.
+///  3. A request to a JavaScript asset in the form of `/packages/foo/bar.js`.
+///     These are looked up relative to the correct package root of the
+///     dart_tool directory.
+///  4. A request to a Dart asset in the form of `/packages/foo/bar.dart` for
+///     sourcemaps. These either need to be looked up from the application lib
+///     directory (if the package is the same), or found in the .packages file.
+///  5. A request for a specific dart asset such as `stack_trace_mapper.js` or
+///     `dart_sdk.js`. These have fixed locations determined by [artifacts].
+///  6. A request to `/` which is translated into `index.html`.
+class WebAssetServer {
+  WebAssetServer(this.flutterProject, this.target, this.ipv6);
+
+  /// The flutter project corresponding to this application.
+  final FlutterProject flutterProject;
+
+  /// The entrypoint we have compiled for.
+  final String target;
+
+  /// Whether to serve from ipv6 localhost.
+  final bool ipv6;
+
+  HttpServer _server;
+  Map<String, Uri> _packages;
+
+  /// The port being served, or null if not initialized.
+  int get port => _server?.port;
+
+  /// Initialize the server.
+  ///
+  /// Throws a [StateError] if called multiple times.
+  Future<void> initialize() async {
+    if (_server != null) {
+      throw StateError('Already serving.');
+    }
+    _packages = PackageMap(PackageMap.globalPackagesPath).map;
+    _server = await HttpServer.bind(
+        ipv6 ? InternetAddress.loopbackIPv6 : InternetAddress.loopbackIPv4, 0)
+      ..autoCompress = false;
+    _server.listen(_onRequest);
+  }
+
+  /// Clean up the server.
+  Future<void> dispose() {
+    return _server.close();
+  }
+
+  /// An HTTP server which provides JavaScript and web assets to the browser.
+  Future<void> _onRequest(HttpRequest request) async {
+    if (request.method != 'GET') {
+      request.response.statusCode = HttpStatus.forbidden;
+      await request.response.close();
+      return;
+    }
+    final Uri uri = request.uri;
+    if (uri.path == '/') {
+      final File file = flutterProject.directory
+          .childDirectory('web')
+          .childFile('index.html');
+      await _completeRequest(request, file, 'text/html');
+    } else if (uri.path.contains('stack_trace_mapper')) {
+      final File file = fs.file(fs.path.join(
+        artifacts.getArtifactPath(Artifact.engineDartSdkPath),
+        'lib',
+        'dev_compiler',
+        'web',
+        'dart_stack_trace_mapper.js'
+      ));
+      await _completeRequest(request, file, 'text/javascript');
+    } else if (uri.path.contains('require.js')) {
+      final HttpClient client = HttpClient();
+      final HttpClientRequest lookupRequest = await client.getUrl(Uri.parse(
+          'https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js'));
+      final HttpClientResponse lookupResponse = await lookupRequest.close();
```

Comment:
If this throws (e.g. if there is no internet connection), it will bring down the asset server. Consider failing more gracefully.

Reply:
Fixed, this is actually vended from the dart sdk